### ISSUE TITLE
CoAP Client: Continuously retransmits if host is not reachable

### ISSUE THREAD

**Issue by anayajoshi** [shamblett/coap#17]
Hello @shamblett ,
I have set **maxRetransmit: 1** in coap configuration file. 
If I sent ping request or any other get or put request. If My host ip address is not reachable, the coap client continuesly retransmits the commands after time out. It should stop retransmitting after retransmit count reached to maxRetransmit.
Please suggest the solution. Is there any setting to achieve this?

**Comment by shamblett** [comment]
Yes this does seem to be a bug, I'll have a look at this.

**Comment by shamblett** [comment]
OK, I've updated the code to fix this effect, its on branch issue17 of the repo, if you want to test it please update your pubspec.yaml file to get this version of the package, may not be fully there yet but its good enough for testing, my test code -
```
import 'dart:async';
import 'package:coap/coap.dart';
import 'package:test/test.dart';
import 'config/coap_config.dart';

FutureOr main() async {
  test('Transmission', () async {
    // Create a configuration class. Logging levels can be specified in the
    // configuration file.
    final conf = CoapConfig();
    print('ISSUE: max retransmit from configuration is ${conf.maxRetransmit}');

    // Build the request uri, note that the request paths/query parameters can be changed
    // on the request anytime after this initial setup.
    const host = 'google.com';

    final uri = Uri(scheme: 'coap', host: host, port: conf.defaultPort);

    // Create the client.
    // The method we are using creates its own request so we do not
    // need to supply one.
    // The current request is always available from the client.
    final client = CoapClient(uri, conf);

    // Create the request for the get request
    final request = CoapRequest.withType(CoapCode.methodGET);
    request.addUriPath('obs');
    print('ISSUE: max retransmit from request is ${request.maxRetransmit}');

    // You can override the configured maxretransmit setting on a request by request basis
    //request.maxRetransmit = 2;
    print('ISSUE: max retransmit from request is now ${request.maxRetransmit}');
    // Getting responses from the observable resource
    request.responses.listen((CoapResponse response) {
      print('ISSUE: - payload: ${response.payloadString}');
    });

    client.request = request;

    print('ISSUE: - Sending get request to '
        '$host, waiting for responses ....');
    await client.get();
  }, timeout: Timeout.factor(3));
}
```
Two things to note, retransmission timeout is only for confirmable messages(which ping is of course) and the global configuration setting can be overridden on a request by request basis as the test above shows -
```
// You can override the configured maxretransmit setting on a request by request basis
    //request.maxRetransmit = 2;
```

**Comment by anayajoshi** [comment]
> OK, I've updated the code to fix this effect, its on branch issue17 of the repo, if you want to test it please update your pubspec.yaml file to get this version of the package, may not be fully there yet but its good enough for testing, my test code -

I didn't understand this statement. How can I update my pubspec.yaml file to get this version of the package? I am using coap: ^2.0.1 as dependancy in my pubspec.yaml ,  is there any specific syntax changes?

> Two things to note, retransmission timeout is only for confirmable messages(which ping is of course) and the global configuration setting can be overridden on a request by request basis as the test above shows -

As per your suggestion, I set **request.maxRetransmit = 2;** to my put request, but it not works as expected. I continuously sends retransmit command even if host not reachable. It should stop after 2 transmits.

Am I going wrong anywhere?

**Comment by shamblett** [comment]
```
 is there any specific syntax changes?
```
Yes , check the pubspec docs to find out how you can use a git repo with a specific branch as a package reference rather than a version number. There is no point continuingt with 2.0.1 as we already know it won't work.

**Comment by anayajoshi** [comment]
ok, I will check it accordingly.
Still I am waiting for next version update of coap library.

**Comment by shamblett** [comment]
Package updated and re published at version 2.0.2, please re test.

**Comment by anayajoshi** [comment]
ok, I have updated package. This issue is solved.
Thank you so much.
